File: src/com/company/tree/IsSymmetric.java
package com.company.tree;

public class IsSymmetric {

    public static void main(String[] args) {
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(2);
        root.left.left = new TreeNode(3);
        root.left.right = new TreeNode(4);
        root.right.left = new TreeNode(5);
        root.right.right = new TreeNode(3);


        System.out.println(isSymmetric(root));
    }


    public static boolean isSymmetric(TreeNode root) {
        return root == null || isSymmetricHelper(root.left, root.right);
    }

    private static boolean isSymmetricHelper(TreeNode left, TreeNode right) {
        if (left == null || right == null) {
            return left == right;
        }
        if (left.value != right.value) {
            return false;
        }

        return isSymmetricHelper(left.left, right.right) && isSymmetricHelper(left.right, right.left);
    }


}

File: src/com/company/tree/IsSymmetric.kt
package com.company.tree

fun isSymmetric(root: TreeNode?): Boolean {
    return root == null || isSymmetricHelper(root.left, root.right)
}

private fun isSymmetricHelper(left: TreeNode?, right: TreeNode?): Boolean {
    if (left == null || right == null) {
        return left === right
    }
    return if (left.`val` == right.`val`) {
        isSymmetricHelper(left.left, right.right) && isSymmetricHelper(left.right, right.left)
    } else {
        false
    }
}

fun main() {
    val root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(2)
    root.left?.left = TreeNode(3)
    root.left?.right = TreeNode(4)
    root.right?.left = TreeNode(5)
    root.right?.right = TreeNode(3)
    println(isSymmetric(root))
}
